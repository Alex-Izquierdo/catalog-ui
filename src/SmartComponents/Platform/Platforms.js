import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Route, Switch } from 'react-router-dom';
import { Section } from '@red-hat-insights/insights-frontend-components';
import ContentGallery from '../../SmartComponents/ContentGallery/ContentGallery';
import PlatformCard from '../../PresentationalComponents/Platform/PlatformCard';
import PlatformToolbar from '../../PresentationalComponents/Platform/PlatformToolbar';
import { fetchPlatforms } from '../../redux/Actions/PlatformActions';
import { scrollToTop } from '../../Helpers/Shared/helpers';
import Platform from './Platform';

const platformsRoutes = {
  platforms: '',
  detail: '/detail/:id'
};

class Platforms extends Component {
    state = {
      filterValue: '',
      isOpen: false
    };

    fetchData = () => this.props.fetchPlatforms();

    componentDidMount() {
      this.fetchData();
      scrollToTop();
    }

    handleFilterChange = filterValue => this.setState({ filterValue })

    renderPlatforms = () => {
      const filteredItems = {
        items: this.props.platforms
        .filter(({ name }) => name.toLowerCase().includes(this.state.filterValue.toLowerCase()))
        .map((item) => <PlatformCard key={ item.id } { ...item } />),
        isLoading: this.props.isLoading
      };
      return (
        <Fragment>
          <PlatformToolbar onFilterChange={ this.handleFilterChange } searchValue={ this.state.filterValue } title="Platforms" />
          <ContentGallery { ...filteredItems } />
        </Fragment>
      );
    }

    render() {
      return (
        <Section>
          <Switch>
            <Route path={ `/platforms${platformsRoutes.detail}` } component={ Platform } />
            <Route exact path={ `/platforms${platformsRoutes.platforms}` } render={ () => this.renderPlatforms() } />
          </Switch>
        </Section>
      );
    }
}

const mapStateToProps = ({ platformReducer: { platforms, isLoading, filterValue }}) => ({
  platforms,
  isLoading,
  searchFilter: filterValue
});

const mapDispatchToProps = dispatch => {
  return {
    fetchPlatforms: apiProps => dispatch(fetchPlatforms(apiProps))
  };
};

Platforms.propTypes = {
  filteredItems: PropTypes.array,
  platforms: PropTypes.array,
  isLoading: PropTypes.bool,
  searchFilter: PropTypes.string,
  showModal: PropTypes.func,
  hideModal: PropTypes.func,
  history: PropTypes.object,
  fetchPlatforms: PropTypes.func
};

export default connect(mapStateToProps, mapDispatchToProps)(Platforms);
