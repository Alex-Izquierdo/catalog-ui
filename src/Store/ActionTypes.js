export const FETCH_CATALOG_ITEMS = 'FETCH_CATALOG_ITEMS';
export const FETCH_CATALOG_ITEM = 'FETCH_CATALOG_ITEM';
export const FILTER_CATALOG_ITEMS = 'FILTER_CATALOG_ITEMS';

export const FETCH_PROVIDER_DATA = 'FETCH_PROVIDER_DATA';
export const ADD_PROVIDER = 'ADD_PROVIDER';

export const FETCH_SERVICE_PLANS = 'FETCH_SERVICE_PLANS';
export const FETCH_SERVICE_PLAN_PARAMETERS = 'FETCH_SERVICE_PLAN_PARAMETERS';
export const UPDATE_SERVICE_DATA = 'UPDATE_SERVICE_DATA';
export const SUBMIT_SERVICE_ORDER = 'SUBMIT_SERVICE_ORDER';
export const SET_SERVICE_PLAN = 'SET_SERVICE_PLAN';

export const SHOW_MODAL = 'SHOW_MODAL';
export const HIDE_MODAL = 'HIDE_MODAL';