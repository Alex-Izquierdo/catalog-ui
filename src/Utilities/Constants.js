export const DEFAULT_PAGE_SIZE = 50;

const topoApiHost = process.env.TOPO_API_HOST || '';
const topoApiPort = process.env.TOPO_API_PORT || '443';
const topoBasePath = process.env.TOPO_BASE_PATH;
const topoApiVersion = process.env.TOPO_API_VERSION || '/api/v0.0';
export const SOURCES_API_BASE = `https://${topoApiHost}:${topoApiPort}${topoBasePath}${topoApiVersion.replace(/\/+$/, '')}`;

const apiHost = process.env.API_HOST || 'localhost';
const apiPort = process.env.API_PORT || '5000';
const basePath = process.env.BASE_PATH || '';
const apiVersion = process.env.API_VERSION || '/api/v0.0';
export const SERVICE_PORTAL_API_BASE = `https://${apiHost}:${apiPort}${basePath}${apiVersion.replace(/\/+$/, '')}`;
