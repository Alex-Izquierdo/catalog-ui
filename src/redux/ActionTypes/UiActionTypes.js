const prefix = '@@catalog/ui/';

export const TOGGLE_EDIT_SERVICE_PORTAL_ITEM = `${prefix}toggle-edit`;
