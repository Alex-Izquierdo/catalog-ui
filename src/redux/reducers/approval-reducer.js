import { SET_LOADING_STATE } from '../action-types';
import { ASYNC_ACTIONS } from '../action-types/approval-action-types';

export const approvalInitialState = {
  isFetching: false,
  isResolving: false,
  workflows: [],
  resolvedWorkflows: []
};

const setLoadingState = (state, { payload = true }) => ({
  ...state,
  isFetching: payload
});
const setWorkflows = (state, { payload }) => ({
  ...state,
  isFetching: false,
  workflows: payload
});
const setResolvingState = (state, { payload = true }) => ({
  ...state,
  isResolving: payload
});
const setResolvedWorkflows = (state, { payload }) => ({
  ...state,
  isResolving: false,
  resolvedWorkflows: payload
});

export default {
  [ASYNC_ACTIONS.FETCH_WORKFLOWS_PENDING]: setLoadingState,
  [ASYNC_ACTIONS.FETCH_WORKFLOWS_FULFILLED]: setWorkflows,
  [ASYNC_ACTIONS.RESOLVE_WORKFLOWS_PENDING]: setResolvingState,
  [ASYNC_ACTIONS.RESOLVE_WORKFLOWS_FULFILLED]: setResolvedWorkflows,
  [SET_LOADING_STATE]: setLoadingState
};
