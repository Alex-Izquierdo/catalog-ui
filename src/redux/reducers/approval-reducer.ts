import { Workflow } from '@redhat-cloud-services/approval-client';
import { SET_LOADING_STATE } from '../action-types';
import { ASYNC_ACTIONS } from '../action-types/approval-action-types';
import { StateFromReducersMapObject } from 'redux';
import { ReduxAction, AnyObject } from '../../types/common-types';

export interface ApprovalReducerState extends AnyObject {
  isFetching: boolean;
  isResolving: boolean;
  workflows: Workflow[];
  resolvedWorkflows: Workflow[];
}

export type ApprovalReducerActionHandler = (
  state: StateFromReducersMapObject<ApprovalReducerState>,
  action: ReduxAction
) => ApprovalReducerState;

export const approvalInitialState = {
  isFetching: false,
  isResolving: false,
  workflows: [],
  resolvedWorkflows: []
};

const setLoadingState: ApprovalReducerActionHandler = (
  state,
  { payload = true }
) => ({
  ...state,
  isFetching: payload
});
const setWorkflows: ApprovalReducerActionHandler = (state, { payload }) => ({
  ...state,
  isFetching: false,
  workflows: payload
});
const setResolvingState: ApprovalReducerActionHandler = (
  state,
  { payload = true }
) => ({
  ...state,
  isResolving: payload
});
const setResolvedWorkflows: ApprovalReducerActionHandler = (
  state,
  { payload }
) => ({
  ...state,
  isResolving: false,
  resolvedWorkflows: payload
});

export default {
  [ASYNC_ACTIONS.FETCH_WORKFLOWS_PENDING]: setLoadingState,
  [ASYNC_ACTIONS.FETCH_WORKFLOWS_FULFILLED]: setWorkflows,
  [ASYNC_ACTIONS.RESOLVE_WORKFLOWS_PENDING]: setResolvingState,
  [ASYNC_ACTIONS.RESOLVE_WORKFLOWS_FULFILLED]: setResolvedWorkflows,
  [SET_LOADING_STATE]: setLoadingState
};
