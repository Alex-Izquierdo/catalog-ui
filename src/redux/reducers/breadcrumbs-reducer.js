import { INITIALIZE_BREADCRUMBS } from '../action-types';

export const initialBreadcrumbsState = {
  fragments: []
};

const initialize = (state, { payload }) => ({ ...state, fragments: payload });

export default {
  [INITIALIZE_BREADCRUMBS]: initialize
};
