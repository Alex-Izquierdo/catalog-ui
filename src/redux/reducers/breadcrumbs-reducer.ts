import { ReactNode } from 'react';
import { StateFromReducersMapObject } from 'redux';
import { INITIALIZE_BREADCRUMBS } from '../action-types';
import { AnyObject, ReduxAction } from '../../types/common-types';

export interface BreadcrumbFragment {
  pathname: string;
  searchParams?: string;
  title?: ReactNode;
}
export interface BreadcrumbsReducerState extends AnyObject {
  fragments: BreadcrumbFragment[];
}

export type BreadcrumbsReducerActionHandler = (
  state: StateFromReducersMapObject<BreadcrumbsReducerState>,
  action: ReduxAction
) => BreadcrumbsReducerState;

export const initialBreadcrumbsState: BreadcrumbsReducerState = {
  fragments: []
};

const initialize: BreadcrumbsReducerActionHandler = (state, { payload }) => ({
  ...state,
  fragments: payload
});

export default {
  [INITIALIZE_BREADCRUMBS]: initialize
};
