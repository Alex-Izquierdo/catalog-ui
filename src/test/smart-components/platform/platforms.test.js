import React from 'react';
import { mount, shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import configureStore from 'redux-mock-store' ;
import thunk from 'redux-thunk';
import promiseMiddleware from 'redux-promise-middleware';
import { MemoryRouter } from 'react-router-dom';
import { notificationsMiddleware } from '@red-hat-insights/insights-frontend-components/components/Notifications';
import { platformInitialState } from '../../../redux/reducers/platformReducer';
import Platforms from '../../../SmartComponents/Platform/Platforms';
import { TOPOLOGICAL_INVENTORY_API_BASE } from '../../../Utilities/Constants';
import { FETCH_PLATFORMS } from '../../../redux/ActionTypes';
import { mockBreacrumbsStore } from '../../redux/redux-helpers';

describe('<Platforms />', () => {
  let initialProps;
  let initialState;
  const middlewares = [ thunk, promiseMiddleware(), notificationsMiddleware() ];
  let mockStore;
  beforeEach(() => {
    mockStore = configureStore(middlewares);
    initialProps = {};
    initialState = {
      platformReducer: {
        ...platformInitialState
      }
    };
  });

  it('should render correctly', () => {
    const wrapper = shallow(<Platforms store={ mockStore(initialState) } />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should mount and fetch platforms data', (done) => {
    const store = mockStore(initialState);
    apiClientMock.get(`${TOPOLOGICAL_INVENTORY_API_BASE}/sources`, mockOnce({ body: { data: [{ id: '1', name: 'foo' }]}}));
    mount(<MemoryRouter><Platforms { ...initialProps } store={ store } /></MemoryRouter>);
    setImmediate(() => {
      const expectedActions = [{
        type: `${FETCH_PLATFORMS}_PENDING`
      }, {
        type: `${FETCH_PLATFORMS}_FULFILLED`,
        payload: [{ id: '1', name: 'foo' }]
      }];
      expect(store.getActions()).toEqual(expectedActions);
      done();
    });
  });

  it('should call filter handler and filter data', (done) => {
    initialState = {
      platformReducer: {
        ...platformInitialState,
        platforms: [{
          id: '1',
          name: 'Foo',
          description: 'desc',
          modified: 'mod'
        }, {
          id: '2',
          name: 'Bar',
          description: 'desc',
          modified: 'mod'
        }, {
          id: '3',
          name: 'baz',
          description: 'desc',
          modified: 'mod'
        }]
      }
    };
    const Provider = mockBreacrumbsStore(initialState, middlewares);
    apiClientMock.get(`${TOPOLOGICAL_INVENTORY_API_BASE}/sources`, mockOnce({ body: { data: [{ id: '1', name: 'foo' }]}}));
    const wrapper = mount(
      <Provider>
        <MemoryRouter initialEntries={ [ '/platforms' ] }><Platforms { ...initialProps } /></MemoryRouter>
      </Provider>
    );

    setImmediate(() => {
      const search = wrapper.find('input');

      search.getDOMNode().value = 'foo';
      search.simulate('change');
      wrapper.update();
      expect(wrapper.find(Platforms).children().instance().state.filterValue).toEqual('foo');
      done();
    });
  });
});
