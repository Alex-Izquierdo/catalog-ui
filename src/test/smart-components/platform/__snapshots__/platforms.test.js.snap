// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Platforms /> should render correctly 1`] = `
<ContextProvider
  value={null}
>
  <Platforms
    fetchPlatforms={[Function]}
    isLoading={false}
    platforms={Array []}
    searchFilter=""
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />
</ContextProvider>
`;
